### PR TITLE
HPCC-23560 Add build-time flag for containerized mode

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
Minikube

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Please review. This is a first step towards:

a) using a dedicated build flag for containerized builds, rather than making a standard package and installing it
b) Reducing the number of places where an environment.xml/environment.conf is processed

File: ecl/hql/hql.hpp [hunk 1]
```diff
@@ -24,7 +24,6 @@
 #endif
 
 #include "hqlatoms.hpp"
-#include "build-config.h"
```
Comment (richardkchapman) [inline ecl/hql/hql.hpp]: These changes are also present in a separate PR, which should be merged first and this one rebased.

File: system/jlib/jptree.cpp [hunk 2]
```diff
@@ -7905,6 +7905,14 @@ jlib_decl IPropertyTree * loadConfiguration(const char * defaultYaml, const char
             displayConfig(config, componentTag);
             exit(0);
         }
+#ifdef _DEBUG
+        else if (strsame(cur, "--hold"))
```
Comment (richardkchapman) [inline system/jlib/jptree.cpp]: This allows any component to be placed into "hold" mode at startup for ease of debugging (you can then connect to the pod manually using kubectl exec, and start the process under a debugger).

File: dali/server/daserver.cpp [hunk 3]
```diff
@@ -365,6 +367,14 @@ int main(int argc, char* argv[])
             port = atoi(argv[++i]);
         else if (streq(argv[i],"--rank") || streq(argv[i],"-r"))
             myrank = atoi(argv[++i]);
+#ifdef _DEBUG
+        else if (streq(argv[i],"--hold"))
```
Comment (richardkchapman) [inline dali/server/daserver.cpp]: See comment below re hold mode

Comment (ghalliday) [reply dali/server/daserver.cpp]: I assume this will disappear once dali calls loadConfiguration?

**Review by ghalliday — COMMENTED** [review]
@richardkchapman minor comments, also should be rebased now the other PR has been merged

File: dali/server/daserver.cpp (hunk 3 above)
Comment (richardkchapman) [reply dali/server/daserver.cpp]: Yes

**Comment by richardkchapman** [comment]
@ghalliday I have rebased

File: dockerfiles/platform-core/Dockerfile [hunk 4]
```diff
@@ -29,17 +29,26 @@ RUN apt-get install -y \
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply dockerfiles/platform-core/Dockerfile]: https://track.hpccsystems.com/browse/HPCC-23532

File: common/workunit/workunit.cpp [hunk 5]
```diff
@@ -7788,7 +7783,8 @@ IConstWUClusterInfo* getTargetClusterInfo(IPropertyTree *environment, IPropertyT
         if (thorName) 
         {
             xpath.clear().appendf("Software/ThorCluster[@name=\"%s\"]", thorName);
-            thors.append(*environment->getPropTree(xpath.str()));
+            if (environment->hasProp(xpath.str()))
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply common/workunit/workunit.cpp]: If the efficiency of this code matters, we are doing something wrong

[the remark this answers is not in the document]
Comment (ghalliday) [reply common/workunit/workunit.cpp]: That may be true, but still probably better to not change it so it is worse.
@jakesmith will this impose any extra load on dali, or will it all be cached locally?  I assume the latter.

[the remark this answers is not in the document]
Comment (richardkchapman) [reply common/workunit/workunit.cpp]: Better still would be to redesign all of this code so we don't retrieve information that we really don't need (I don't need to know anything about the thor configuration to find the name of the eclagent queue that eclcc wants to push onto, which was what I was trying to get to work when I made this change...

**Pull request merged**